### PR TITLE
Implement object.pattern(regex, schema)

### PR DESCRIPTION
**Pull request by EthanWaite**
This implements the object.pattern() method from Joi, which lets you validate an object with unknown keys.
https://github.com/hapijs/joi/blob/master/API.md#objectpatternregex-schema

### CODE REVIEW

File: lib/rules/object.js [hunk 1]
```diff
@@ -93,6 +93,54 @@ class Keys extends Rule {
...
+    compile(params) {
+        this._keyRegex = new RegExp(params.args[0]);
+        this._valueSchema = params.args[1];
+    }
+
+    validate(params, callback) {
+        const value = params.value;
+        if (value === undefined || value === null) {
+            return callback(this.error(params));
+        }
```
Comment (connor4312) [inline lib/rules/object.js]: I don't think we should bail here, the consumer might allow null or undefined values to be set as the property.

```
.pattern(/./, Jo.number().allow(null))
```

They can add `.required()` if they don't want to allow the values to be undefined

File: lib/rules/object.js [hunk 2]
```diff
@@ -93,6 +93,54 @@ class Keys extends Rule {
...
+
+        const todo = Object.keys(value).map((key) => {
```
Comment (connor4312) [inline lib/rules/object.js]: You can use `async.map(Object.keys(value), key => {` here

File: lib/rules/object.js [hunk 3]
```diff
@@ -93,6 +93,54 @@ class Keys extends Rule {
...
+            const options = clone(params.options);
```
Comment (connor4312) [inline lib/rules/object.js]: Probably only need to clone the options once, you can reuse the cloned object

File: lib/rules/object.js (hunk 2 above)
Comment (EthanWaite) [inline lib/rules/object.js]: @connor4312 Jojen doesn't use the library "async" - it uses https://github.com/WatchBeam/jojen/blob/master/lib/util.js#L1

Comment (connor4312) [inline lib/rules/object.js]: Ah, right, ok. It's fine then, though you can add a map function if you want 😛

**Comment by connor4312** [comment]
rmed the comment about reusing, doesn't work with parallel async validations

**Pull request merged**